### PR TITLE
[ADP-3359] Report unaccepted-era transactions as HTTP 403 errors

### PR DESCRIPTION
**Pull request by paolino**
- [x] Remove an `error` call when not-in-a-recent-era  tx is posted
- [x] Add a 403 HTTP error in case the exception above is triggered

ADP-3359

### CODE REVIEW

File: lib/network-layer/src/Cardano/Wallet/Network.hs [hunk 1]
```diff
@@ -275,14 +276,20 @@ mapChainFollower fpoint12 fpoint21 ftip fblocks cf =
...
+    | ErrPostTxEraUnsupported AnyCardanoEra
     deriving (Generic, Show, Eq)
 
 instance ToText ErrPostTx where
     toText = \case
         ErrPostTxValidationError msg -> msg
         ErrPostTxMempoolFull ->
             "mempool was full and refused posted transaction"
+        ErrPostTxEraUnsupported era ->
+            "submitted transaction is not compatible with the current era: "
```
Comment (paweljakubas) [inline lib/network-layer/src/Cardano/Wallet/Network.hs]: I would rephrase that ... it is not supported with Babbage and Conway at this moment. So tx constructed within Alozno, Mary ,... will not be supported by us

**Review by paweljakubas — CHANGES_REQUESTED** [review]
When adding `UnsupportedEraInTxSubmission` you need to add its counterparts in swagger, so something like : 
```
x-errUnsupportedEraInTxSubmission: &errUnsupportedEraInTxSubmission                                                                                                                                                                  
   <<: *responsesErr                                                                                                                                                                                           
   title: unsupported_era_in_tx_submission                                                                                                                                                                                  
   properties:                                                                                                                                                                                                               message:                                                                                                                                                                                                  
       type: string                                                                                                                                                                                            
       description: |                                                                                                                                                                                          
         Happens when transaction was constructed before Babbage era                                                                                                                                                  
     code:                                                                                                                                                                                                     
       type: string                                                                                                                                                                                            
       enum: ['unsupported_era_in_tx_submission']
```
And insert `errUnsupportedEraInTxSubmission ` under proper submit endpoint in 403 hierarchy.
Without it unit test will fail as there is no correspondence between ApiError and swagger errors.

Also, what about adding integration test with Alonzo tx as a cbor and showing that the endpoint endup with 403 and proper error msg?

File: lib/api/src/Cardano/Wallet/Api/Types/Error.hs [hunk 2]
```diff
@@ -217,6 +217,7 @@ data ApiErrorInfo
     | BlockHeaderNotFound
     | TranslationByronTxOutInContext
     | BalanceTxInlinePlutusV3ScriptNotSupportedInBabbage
+    | UnsupportedEraInTxSubmission
```
Comment (jonathanknowles) [inline lib/api/src/Cardano/Wallet/Api/Types/Error.hs]: Two suggestions:

1. Perhaps we can **_shorten_** the name to `UnsupportedEra`?
    Justification: The context of the error is (presumably) already known the to caller. Additionally, if in future we wish to return a similar error from another endpoint, we can just reuse this one. Otherwise, we might end up needing a family of similarly-named errors `UnsupportedEraIn{Foo,Bar,Baz,..}`.

2. Perhaps we can make this into a **_rich error type_**?

Something like this:
```suggestion
    | UnsupportedEra !ApiErrorUnsupportedEra
        
data ApiErrorUnsupportedEra = ApiErrorUnsupportedEra
    { unsupportedEra :: !ApiEra
    -- ^ The unsupported era (as specified by the caller).
    , supportedEras :: !(Set ApiEra)
    -- ^ The set of eras that we currently support.
    }
    deriving (Data, Eq, Generic, Show, Typeable)
    deriving (FromJSON, ToJSON) via DefaultRecord ApiErrorUnsupportedEra
    deriving anyclass NFData
```

File: lib/network-layer/src/Cardano/Wallet/Network.hs [hunk 3]
```diff
@@ -275,14 +276,20 @@ mapChainFollower fpoint12 fpoint21 ftip fblocks cf =
...
+                <> T.pack (show era)
```
Comment (jonathanknowles) [inline lib/network-layer/src/Cardano/Wallet/Network.hs]: Suggestion:

Instead of doing string interpolation here (which is awkward to test), I would propose to use a **_fixed message string_** that defers to a rich error object. (See [this comment](https://github.com/cardano-foundation/cardano-wallet/pull/4595#discussion_r1603193349)!)

File: lib/primitive/lib/Cardano/Wallet/Primitive/Ledger/Shelley.hs [hunk 4]
```diff
@@ -1018,21 +1021,21 @@ rewardAccountFromAddress (W.Address bytes) = refToAccount . ref =<< parseAddr by
     refToAccount (SL.StakeRefPtr _) = Nothing
     refToAccount SL.StakeRefNull = Nothing
 
+newtype UnsealedTxException = UnsealedTxInUnsupportedEra AnyCardanoEra
```
Comment (jonathanknowles) [inline lib/primitive/lib/Cardano/Wallet/Primitive/Ledger/Shelley.hs]:
```suggestion
newtype UnsealedTxException = UnsealedTxInUnsupportedEra AnyCardanoEra

```

File: lib/api/src/Cardano/Wallet/Api/Types/Error.hs (hunk 2 above)
Comment (jonathanknowles) [reply lib/api/src/Cardano/Wallet/Api/Types/Error.hs]: It turns out we already have a **_similar_** API error (but for a very different scenario):

https://github.com/cardano-foundation/cardano-wallet/blob/47f4ccc0bab51c323be128fb7adf1cda0be80dbc/lib/api/src/Cardano/Wallet/Api/Types/Error.hs#L274-L281

File: lib/api/src/Cardano/Wallet/Api/Http/Server/Error.hs [hunk 5]
```diff
@@ -626,18 +628,35 @@ instance IsServerError ErrRemoveTx where
...
-                ]
+            apiError
+                err425
+                    { errBody = "Mempool is full, please try resubmitting again later."
+                    }
+                MempoolIsFull
+                $ mconcat
+                    [ "The submitted transaction was rejected by the Cardano node "
+                    , "because its mempool was full."
+                    ]
```
Comment (jonathanknowles) [inline lib/api/src/Cardano/Wallet/Api/Http/Server/Error.hs]: **Request**: limit line length to 80 characters here, as before.

Perhaps we could just revert these lines (since they're unrelated to the PR)?

Or alternatively:
```suggestion
            apiError
                err425
                    { errBody =
                        "Mempool is full, please try resubmitting again later."
                    }
                MempoolIsFull
                $ mconcat
                    [ "The submitted transaction was rejected by the Cardano "
                    , "node because its mempool was full."
                    ]
```

File: lib/api/src/Cardano/Wallet/Api/Http/Server/Error.hs [hunk 6]
```diff
@@ -626,18 +628,35 @@ instance IsServerError ErrRemoveTx where
...
+          where
+            error' =
+                UnsupportedEra
+                    $ ApiErrorUnsupportedEra
+                        (toApiEra unsupported)
+                    $ Set.fromList
+                        ( toApiEra
+                            . Write.toAnyCardanoEra
+                            <$> [minBound .. maxBound]
+                        )
```
Comment (jonathanknowles) [inline lib/api/src/Cardano/Wallet/Api/Http/Server/Error.hs]: IMHO it's not very easy for readers to see that:
- this is constructing a record with two fields (due to different indent levels)
- the second field is the _set of all supported recent eras_

**Suggestion**:

Rewrite this so that fields are named explicitly:

```suggestion
            flip (apiError err403) (toText e) $ UnsupportedEra
                ApiErrorUnsupportedEra
                    { unsupportedEra = toApiEra unsupported
                    , supportedEras =
                        Set.fromList $
                        map (toApiEra . Write.toAnyCardanoEra)
                            [minBound .. maxBound]
                    }
```

This also saves a bit of space, and avoids having to define `error'`.

File: lib/api/src/Cardano/Wallet/Api/Types/Error.hs [hunk 7]
```diff
@@ -235,6 +240,16 @@ apiErrorInfoOptions = defaultSumTypeOptions
...
 
+data ApiErrorUnsupportedEra = ApiErrorUnsupportedEra
+    { unsupportedEra :: !ApiEra
+    -- ^ The unsupported era (as specified by the caller).
+    , supportedEras :: !(Set ApiEra)
+    -- ^ The set of eras that we currently support.
+    }
+    deriving (Data, Eq, Generic, Show, Typeable)
+    deriving (FromJSON, ToJSON) via DefaultRecord ApiErrorUnsupportedEra
+    deriving anyclass NFData
```
Comment (jonathanknowles) [inline lib/api/src/Cardano/Wallet/Api/Types/Error.hs:251]: Very nice! 👍🏻

File: lib/network-layer/src/Cardano/Wallet/Network.hs [hunk 8]
```diff
@@ -106,6 +106,7 @@ import qualified Internal.Cardano.Write.Tx as Write
 {-----------------------------------------------------------------------------
     NetworkLayer
 ------------------------------------------------------------------------------}
+
```
Comment (jonathanknowles) [inline lib/network-layer/src/Cardano/Wallet/Network.hs:109]: 👍🏻

File: lib/network-layer/src/Cardano/Wallet/Network/Implementation.hs [hunk 9]
```diff
@@ -630,12 +631,16 @@ withNodeNetworkLayerBase
...
-                        $ unsealShelleyTx preferredEra tx
-            liftIO (send txSubmissionQueue cmd) >>= \case
-                SubmitSuccess -> pure ()
-                SubmitFail e -> throwE $ ErrPostTxValidationError $ T.pack $ show e
+            case unsealShelleyTx preferredEra tx of
+                Left (UnsealedTxInUnsupportedEra era) ->
+                    throwE $ ErrPostTxEraUnsupported era
+                Right tx' -> do
+                    let cmd =
+                            CmdSubmitTx . toConsensusGenTx $ tx'
```
Comment (jonathanknowles) [inline lib/network-layer/src/Cardano/Wallet/Network/Implementation.hs]: **Suggestion:**

```suggestion
                    let cmd = CmdSubmitTx . toConsensusGenTx $ tx'
```

File: lib/primitive/lib/Cardano/Wallet/Primitive/Ledger/Shelley.hs [hunk 10]
```diff
@@ -1018,21 +1021,22 @@ rewardAccountFromAddress (W.Address bytes) = refToAccount . ref =<< parseAddr by
...
```
Comment (jonathanknowles) [inline lib/primitive/lib/Cardano/Wallet/Primitive/Ledger/Shelley.hs]: **Suggestion:**

Since this exception pertains to the `unsealTx` operation, and since it is only returned if it is **_not_** possible to unseal a transaction:

```suggestion
newtype UnsealTxException = UnsealTxInUnsupportedEra AnyCardanoEra
```

File: specifications/api/swagger.yaml [hunk 11]
```diff
@@ -5557,6 +5557,26 @@ x-errBlockHeaderNotFound: &errBlockHeaderNotFound
...
+    info:
+      type: object
+      required:
+        - unsupported_era
+        - supported_eras
+      properties:
+          unsupported_era: *ApiEra
+          supported_eras:
+            type: array
+            items: *ApiEra
```
Comment (jonathanknowles) [inline specifications/api/swagger.yaml:5578]: 👍🏻

**Review by jonathanknowles — COMMENTED** [review]
Hi @paolino 
Many thanks for making this PR.
I've left some suggestions (and one small request)!

**Review by jonathanknowles — APPROVED** [review]
LGTM!

**Review by paweljakubas — APPROVED** [review]
nice! I have added integration tests showing that your changes handle tx constructed in previous eras as it is expected .
